src/MultiFormats/Cryptography/KeccakManaged.cs

```diff
@@ -0,0 +1,374 @@
+// This is a copy of https://bitbucket.org/jdluzen/sha3/raw/d1fd55dc225d18a7fb61515b62d3c8f164d2e788/SHA3Managed/SHA3Managed.cs
+
+using System;
+
+namespace MultiFormats.Cryptography
+{
+    internal sealed class KeccakManaged : Keccak
+    {
+        public KeccakManaged(int hashBitLength)
+            : base(hashBitLength) { }
+
+        protected override void HashCore(byte[] array, int ibStart, int cbSize)
+        {
+            base.HashCore(array, ibStart, cbSize);
+            if (cbSize == 0)
+            {
+                return;
+            }
+            
+            var sizeInBytes = SizeInBytes;
+            if (Buffer == null)
+            {
+                Buffer = new byte[sizeInBytes];
+            }
+            
+            var stride = sizeInBytes >> 3;
+            var utemps = new ulong[stride];
+            if (BuffLength == sizeInBytes)
+            {
+                throw new Exception("Unexpected error, the internal buffer is full");
+            }
+            
+            AddToBuffer(array, ref ibStart, ref cbSize);
+            if (BuffLength == sizeInBytes) //buffer full
+            {
+                System.Buffer.BlockCopy(Buffer, 0, utemps, 0, sizeInBytes);
+                KeccakF(utemps, stride);
+                BuffLength = 0;
+            }
+
+            for (; cbSize >= sizeInBytes; cbSize -= sizeInBytes, ibStart += sizeInBytes)
+            {
+                System.Buffer.BlockCopy(array, ibStart, utemps, 0, sizeInBytes);
+                KeccakF(utemps, stride);
+            }
+
+            if (cbSize <= 0)
+            {
+                return;
+            }
+            
+            System.Buffer.BlockCopy(array, ibStart, Buffer, BuffLength, cbSize);
+            BuffLength += cbSize;
+        }
+
+        protected override byte[] HashFinal()
+        {
+            var sizeInBytes = SizeInBytes;
+            var outb = new byte[HashByteLength];
+
+            //    padding
+            if (Buffer == null)
+            {
+                Buffer = new byte[sizeInBytes];
+            }
+            else
+            {
+                Array.Clear(Buffer, BuffLength, sizeInBytes - BuffLength);
+            }
+            
+            Buffer[BuffLength++] = 1;
+            Buffer[sizeInBytes - 1] |= 0x80;
+            var stride = sizeInBytes >> 3;
+            var utemps = new ulong[stride];
+            System.Buffer.BlockCopy(Buffer, 0, utemps, 0, sizeInBytes);
+            KeccakF(utemps, stride);
+            System.Buffer.BlockCopy(state, 0, outb, 0, HashByteLength);
+            return outb;
+        }
+
+        private void KeccakF(ulong[] inb, int laneCount)
+        {
+            while (--laneCount >= 0)
+            {
+                state[laneCount] ^= inb[laneCount];
+            }
+            
+            ulong Aba, Abe, Abi, Abo, Abu;
+            ulong Aga, Age, Agi, Ago, Agu;
+            ulong Aka, Ake, Aki, Ako, Aku;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Declare 'Aku' in a separate statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)